crates/worker/src/partition/state_machine/tests/idempotency.rs

```diff
@@ -142,6 +142,126 @@ async fn start_idempotent_invocation() {
+                request_id,
+            }),
+            idempotency_key: Some(idempotency_key),
+            completion_retention_time: Some(retention),
+            ..ServiceInvocation::mock()
+        }))
+        .await;
+    assert_that!(
+        actions,
+        contains(pat!(Action::Invoke {
+            invocation_id: eq(invocation_id),
+            invoke_input_journal: pat!(InvokeInputJournal::CachedJournal(_, _))
+        }))
+    );
+
+    // Assert idempotency key mapping exists
+    let mut txn = state_machine.storage().transaction();
+    assert_that!(
+        txn.get_idempotency_metadata(&idempotency_id)
+            .await
+            .unwrap()
+            .unwrap(),
+        pat!(IdempotencyMetadata {
+            invocation_id: eq(invocation_id),
+        })
+    );
+    txn.commit().await.unwrap();
+
+    // Send output, then end
+    let response_bytes = Bytes::from_static(b"123");
+    let actions = state_machine
+        .apply_multiple([
+            Command::InvokerEffect(InvokerEffect {
+                invocation_id,
+                kind: InvokerEffectKind::JournalEntry {
+                    entry_index: 1,
+                    entry: ProtobufRawEntryCodec::serialize_enriched(Entry::output(
+                        EntryResult::Success(response_bytes.clone()),
+                    )),
+                },
+            }),
+            Command::InvokerEffect(InvokerEffect {
+                invocation_id,
+                kind: InvokerEffectKind::End,
+            }),
+        ])
+        .await;
+
+    // Assert response and timeout
+    assert_that!(
+        actions,
+        all!(
+            contains(pat!(Action::IngressResponse(pat!(
+                IngressResponseEnvelope {
+                    target_node: eq(node_id),
+                    inner: pat!(ingress::InvocationResponse {
+                        request_id: eq(request_id),
+                        invocation_id: some(eq(invocation_id)),
+                        response: eq(IngressResponseResult::Success(
+                            invocation_target.clone(),
+                            response_bytes.clone()
+                        ))
+                    })
+                }
+            )))),
+            not(contains(pat!(Action::ScheduleInvocationStatusCleanup {
+                invocation_id: eq(invocation_id)
+            })))
+        )
+    );
+
+    // InvocationStatus contains completed
+    let invocation_status = state_machine
+        .storage()
+        .transaction()
+        .get_invocation_status(&invocation_id)
+        .await
+        .unwrap();
+    let_assert!(InvocationStatus::Completed(completed_invocation) = invocation_status);
+    assert_eq!(
+        completed_invocation.response_result,
+        ResponseResult::Success(response_bytes)
+    );
+    assert!(unsafe { completed_invocation.timestamps.completed_transition_time() }.is_some());
+    assert_eq!(completed_invocation.completion_retention, retention);
```

Comment:
Should we also test that the clean up happens? If you start the test with `start_paused = true`, then tokio will advance the time to the next event if nothing else can run. That way we can simulate the hourly cleanup in basically no time. Here are a few more details https://docs.rs/tokio/latest/tokio/time/fn.pause.html.

Reply:
I think what I would rather do is have such test in the cleaner, to check if the PurgeInvocation commands are produced to bifrost.